src/components/street-segment.js

```diff
@@ -0,0 +1,424 @@
+        generatedComponentList.push(componentName);
+      }
+    }
+    this.generatedComponents = generatedComponentList;
+  },
+  update: function (oldData) {
+    const data = this.data;
+    const dataDiff = AFRAME.utils.diff(oldData, data);
+    // if oldData is same as current data, then don't update
+    if (AFRAME.utils.deepEqual(oldData, data)) {
+      return;
+    }
+    // regenerate components if only type has changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('type')
+    ) {
+      let typeObject = this.types[this.data.type];
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      this.remove();
+      this.createGeneratedComponentsFromType(typeObject); // add components for this type
+      this.updateSurfaceFromType(typeObject); // update surface color, surface, level
+    }
+    // propagate change of direction to generated components is solo changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('direction')
+    ) {
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      for (const componentName of this.generatedComponents) {
+        this.el.setAttribute(componentName, 'direction', this.data.direction);
+      }
+    }
+    // propagate change of length to generated components is solo changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('length')
+    ) {
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      for (const componentName of this.generatedComponents) {
+        this.el.setAttribute(componentName, 'length', this.data.length);
+      }
+    }
+    this.clearMesh();
+    this.height = this.calculateHeight(data.level);
+    this.tempXPosition = this.el.getAttribute('position').x;
+    this.el.setAttribute('position', { x: this.tempXPosition, y: this.height });
+    this.generateMesh(data);
+    // if width was changed, trigger re-justification of all street-segments by the managed-street
+    if (Object.keys(dataDiff).includes('width')) {
+      this.el.parentNode.components['managed-street'].applyJustification();
+    }
+  },
+  // for streetmix elevation number values of -1, 0, 1, 2, calculate heightLevel in three.js meters units
+  calculateHeight: function (elevationLevel) {
+    const stepLevel = 0.15;
+    if (elevationLevel <= 0) {
+      return stepLevel;
+    }
+    return stepLevel * (elevationLevel + 1);
+  },
+  clearMesh: function () {
+    // remove the geometry from the entity
+    this.el.removeAttribute('geometry');
+    this.el.removeAttribute('material');
+  },
+  remove: function () {
+    this.clearMesh();
+
+    this.generatedComponents.forEach((componentName) => {
+      this.el.removeAttribute(componentName);
+    });
```

Comment:
`this.generatedComponents.length = 0;` is good practice, more important when you call remove() yourself in update() in some cases.

Reply:
0220b81bf1a7de20ca0d694d7bf497187848f072